_posts/2023-09-06-load-balancer.md

```diff
@@ -0,0 +1,906 @@
+---
+title: "Functional Load Balancer"
+date: 2023-09-06
+header:
+    image: "/images/blog cover.jpg"
+tags: [scala,http4s,cats,fp]
+excerpt: "Building an application load balancer with the help of Scala & cats, featuring efficiency and concurrency "
+---
+
+_by [Anzori (Nika) Ghurtchumelia](https://github.com/ghurtchu)_
+
+## 1. Introduction
+
+"What I cannot build, I do not understand"
+
+~ Richard Feynman
+
+In this blogpost we will make use of `cats.effect.Ref`, `cats.effect.IO` and `http4s` to build an application layer load balancer.
+
+Load balancer usually sits in front of a few servers and forwards the HTTP requests to them based on some algorithm. The goal of load balancer is
+to basically "balance the load" by distributing it to the available backends.
+
+There are different algorithms that can be used to achieve this goal, in this case we're going to be using Round Robin algorithm.
+
+Our load balancer must be featuring the following characteristics:
+- Being effective in distributing requests to the backends (Round Robin Algorithm)
+- Making sure that requests are forwarded only to the available backends (Reliability)
+- Managing the state safely while the backends die and come back online (Periodic health checks)
+
+In practice, every non-trivial application should have:
+- `domain` - data models which describe the domain
+- `services` - bundle of logic which uses `domain` and other `services` to achieve something
+
+Let's start with Domain Modeling.
```

Comment:
Oh yes, I completely skipped the project structure and build.sbt part, will fix.

Reply:
Agreed, let's try with Scala 3.

Reply:
Ok, I made it, but I can't use `opaque type` for `Url` because `pureconfig` doesn't work well for Scala 3. It fails on the derivation phase, even if I define `given reader: ConfigReader[Url]`, so I believe the compromise is to have `Url` as a new type which extends `AnyVal` and defining `given` instances for `Url` and `HealthCheckInterval`, then `Config(..,..,..) derives ConfigReader` works.

Reply:
Or another idea is to not use value classes which removes the need to define `given ConfigReader` instances for `Urls`, and `HealthCheckInterval`, but then the configuration file itself gets uglier:

```scala
port="8080"
host="localhost"
backends={
 values=[
  "localhost:8080",
  "localhost:8081",
  "localhost:8082",
 ]
}
health-check-interval {
 value: "1"
}
```

So I believe the first approach is better, any ideas?

Reply:
I think the code benefits are greater than the config, because you only configure once. I would say stick to `derives ConfigReader`

Reply:
Yep, I kept it, but with value classes, couldn't find solution with `opaque type`-s unfortunately.